// token-server/build/lib/token/TokenServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tls = require("tls");
const Connection_1 = require("../connection/Connection");
const Status_1 = require("./Status");
const TokenAPI_1 = require("./TokenAPI");
class TokenServer extends TokenAPI_1.TokenAPI {
    constructor(options) {
        super();
        /** Contains all active connections */
        this.connections = [];
        this.options = options;
        this.server = tls.createServer(this.options);
        this.applyListeners();
        this.connect();
    }
    close() {
        if (!this.server.listening) {
            // Server already closed.
            return false;
        }
        this.status = Status_1.Status.CLOSED;
        this.server.close();
        for (const connection of this.connections) {
            connection.remoteClose();
        }
        return true;
    }
    connect() {
        if (this.status >= Status_1.Status.CONNECTING) {
            // Server already connecting or online.
            return false;
        }
        this.hadError = false;
        this.status = Status_1.Status.CONNECTING;
        this.server.listen(this.options);
        return true;
    }
    applyListeners() {
        this.server.on('listening', () => {
            this.status = Status_1.Status.ONLINE;
            this.emit('connect');
        });
        this.server.on('error', (error) => {
            // Set status to failed so we know that it was closed by an error.
            this.hadError = true;
            // Emit server errors.
            this.emit('error', error);
            // Call close so the server emits the "close" event.
            this.server.close();
        });
        this.server.on('close', () => {
            this.status = Status_1.Status.OFFLINE;
            // Emit the close event.
            this.emit('close', this.hadError);
        });
        this.server.on('secureConnection', (socket) => {
            const connection = new Connection_1.Connection(socket);
            this.connections.push(connection);
            connection.once('close', () => {
                const i = this.connections.indexOf(connection);
                this.connections.splice(i, 1);
            });
            connection.on('token', (token) => {
                this.emit('token', token, connection);
            });
        });
    }
}
exports.TokenServer = TokenServer;
