// token-server/build/lib/token/TokenClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tls = require("tls");
const Connection_1 = require("../connection/Connection");
const Status_1 = require("./Status");
const TokenAPI_1 = require("./TokenAPI");
class TokenClient extends TokenAPI_1.TokenAPI {
    constructor(options) {
        super();
        this.connection = null;
        this.options = options;
        this.connect();
    }
    close() {
        if (this.status <= Status_1.Status.CLOSED) {
            // Already closed or offline.
            return false;
        }
        this.status = Status_1.Status.CLOSED;
        this.socket.end();
        this.connection = null;
        return true;
    }
    connect() {
        if (this.status >= Status_1.Status.CONNECTING) {
            // Still connected.
            return false;
        }
        this.hadError = false;
        this.status = Status_1.Status.CONNECTING;
        this.socket = tls.connect(this.options);
        this.connection = null;
        this.applyListeners();
        return true;
    }
    /**
     * Send a token to the server.
     * Returns true if the token was written to the connection.
     * @param token The token to transmit to the server.
     */
    send(token) {
        if (this.connection) {
            return this.connection.send(token);
        }
        return false;
    }
    /**
     * Applies initial event listeners.
     */
    applyListeners() {
        this.socket.on('error', (error) => {
            this.hadError = true;
            // Emit socket errors.
            this.emit('error', error);
        });
        this.socket.on('close', () => {
            this.status = Status_1.Status.OFFLINE;
            // Emit own close event.
            this.emit('close', this.hadError);
        });
        this.socket.on('secureConnect', () => {
            this.updateConnection();
            // The socket is now connected.
            this.status = Status_1.Status.ONLINE;
            this.emit('connect');
        });
    }
    /**
     * Creates a new connection object for the socket.
     */
    updateConnection() {
        const connection = new Connection_1.Connection(this.socket);
        connection.on('token', (token) => {
            // Emit token
            this.emit('token', token, connection);
        });
        connection.on('remoteClose', () => {
            this.emit('remoteClose', connection);
        });
        this.connection = connection;
    }
}
exports.TokenClient = TokenClient;
