// token-server/build/lib/connection/Connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const Message_1 = require("./Message");
const CLOSE_TOKEN = Buffer.from('\\\n');
/**
 * Wrapper for streams that emits "token".
 */
class Connection extends events_1.EventEmitter {
    constructor(socket) {
        super();
        this.buffer = Buffer.allocUnsafe(0);
        this.socket = socket;
        this.applyEvents();
    }
    /**
     * Is true if the underlying socket is not writable.
     */
    get isDead() {
        return !this.socket.writable;
    }
    /**
     * Writes a message to the socket. Returns true if the message was written to the socket.
     * @param data The buffer to transmit.
     */
    send(data) {
        if (this.isDead) {
            return false;
        }
        this.socket.write(Message_1.Message.escape(data));
        return true;
    }
    /**
     * Close the connection.
     */
    close() {
        if (this.isDead) {
            return false;
        }
        this.socket.end();
        return true;
    }
    /**
     * Close the other side of the connection.
     * Returns true if the close token was written to the socket.
     */
    remoteClose() {
        if (this.isDead) {
            return false;
        }
        this.socket.write(CLOSE_TOKEN);
        return true;
    }
    /**
     * Initially apply all events.
     */
    applyEvents() {
        this.socket.on('data', (data) => {
            this.buffer = Buffer.concat([this.buffer, data]);
            this.parse();
        });
        // Bubble up close event.
        this.socket.on('close', () => this.emit('close'));
    }
    /**
     * Parses and emits all tokens from the buffer.
     */
    parse() {
        while (true) {
            const i = this.buffer.indexOf(Message_1.NEWLINE);
            if (i === -1) {
                break;
            }
            // +1 to include the separating newline
            const data = this.buffer.slice(0, i + 1);
            if (data.equals(CLOSE_TOKEN)) {
                this.emit('remoteClose');
            }
            else {
                this.emit('token', Message_1.Message.unescape(data));
            }
            this.buffer = this.buffer.slice(i + 1);
        }
    }
}
exports.Connection = Connection;
